File: Algorithm/quickSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int N;

void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

int devideAndConquer(int arr[], int low, int high) {
    int i = low - 1;
    int pivot = arr[high];

    for(int idx = low; idx <= high; idx++) {
        if (arr[idx] < pivot) {
            ++i;
            swap(&arr[i], &arr[idx]);
        }
    }

    swap(&arr[i + 1], &arr[high]);

    return (i + 1);
}

void quickSort(int arr[], int low, int high) {
    if (low < high) {
        int i = devideAndConquer(arr, low, high);
        quickSort(arr, low, i - 1);
        quickSort(arr, i + 1, high);
    }
}

void printArray(int arr[], int size) {
    for (int i = 0; i < size; i++) {
        cout << arr[i] << ' ';
    }
}

int main() {
    int arr[10] = {10, 6, 2, 8, 5, 9, 1, 3, 7, 4};
    int N = 10;
    quickSort(arr, 0, N-1);
    printArray(arr, N);
}

File: Algorithm/Sorting/quickSort.py
def devide_and_conquer(arr, low, high):
    i = low - 1
    pivot = arr[high]

    for idx in range(low, high):
        if arr[idx] < pivot:
            i += 1
            arr[i], arr[idx] = arr[idx], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]

    return i + 1


def quick_sort(arr, low, high):

    if low < high:
        idx = devide_and_conquer(arr, low, high)
        quick_sort(arr, low, idx - 1)
        quick_sort(arr, idx + 1, high)

test_case = [10,5,3,8,6,9,1]
quick_sort(test_case, 0, len(test_case)-1)
print(test_case)

def quicksort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quicksort(left) + middle + quicksort(right)

print(quicksort([3,6,8,10,1,2,1]))
# Prints "[1, 1, 2, 3, 6, 8, 10]"


